src/frontend/src/optimizer/plan_node/batch_hash_join.rs

```diff
@@ -61,16 +59,31 @@ impl BatchHashJoin {
         }
     }
 
-    fn derive_dist(
+    pub(super) fn derive_dist(
         left: &Distribution,
         right: &Distribution,
-        l2o_mapping: &ColIndexMapping,
+        logical: &LogicalJoin,
     ) -> Distribution {
         match (left, right) {
             (Distribution::Single, Distribution::Single) => Distribution::Single,
-            (Distribution::HashShard(_), Distribution::HashShard(_)) => {
-                l2o_mapping.rewrite_provided_distribution(left)
-            }
+            // we can not derive the hash distribution from the side where outer join can generate a
+            // NULL row
+            (Distribution::HashShard(_), Distribution::HashShard(_)) => match logical.join_type() {
+                JoinType::Unspecified => unreachable!(),
+                JoinType::FullOuter => Distribution::SomeShard,
+                JoinType::Inner | JoinType::LeftOuter | JoinType::LeftSemi | JoinType::LeftAnti => {
+                    let l2o = logical
+                        .l2i_col_mapping()
+                        .composite(&logical.i2o_col_mapping());
+                    l2o.rewrite_provided_distribution(left)
+                }
+                JoinType::RightSemi | JoinType::RightAnti | JoinType::RightOuter => {
+                    let r2o = logical
+                        .r2i_col_mapping()
+                        .composite(&logical.i2o_col_mapping());
+                    r2o.rewrite_provided_distribution(right)
+                }
+            },
```

Comment:
PTAL @fuyufjh

Reply:
Make sense. A right join is generated from q15 previously?

Reply:
Not. this PR is trying to solve another problem and add the join key into the `stream_key`. after that, some plan has more columns that are not pruned so that the optimizer can derive more distribution and omit exchange. then the issue is exposed.